expression/util.go

```diff
@@ -66,11 +67,32 @@ func ColumnSubstitute(expr Expression, schema *Schema, newExprs []Expression) Ex
 func datumsToConstants(datums []types.Datum) []Expression {
 	constants := make([]Expression, 0, len(datums))
 	for _, d := range datums {
-		constants = append(constants, &Constant{Value: d})
+		constants = append(constants, &Constant{Value: d, RetType: types.NewFieldType(kindToMysqlType[d.Kind()])})
 	}
 	return constants
 }
 
+var kindToMysqlType = map[byte]byte{
```

Comment:
Looks like it is only used for the test, we can remove the kinds we never used from the map.